File: packages/utils/projections/src/old/SphereProjection.js
/**
 * Copyright (C) 2021 Alibaba Group Holding Limited
 * All rights reserved.
 */

const Projection = require('./Projection')

// eslint-disable-next-line no-unused-vars
const MAX_LATITUDE = 85.05112877980659
const DEG2RAD = Math.PI / 180
const R = 6378137

const defaultProps = { center: [0, 0, 0] }

module.exports = class SphereProjection extends Projection {
	constructor(props) {
		props = { ...defaultProps, ...props }
		super(props)

		this.R = props.R || R

		this.type = 'SphereProjection'

		this.isSphereProjection = true
	}

	get center() {
		return this._center
	}

	set center(center) {
		this._center = center
		this._xyz0 = projRaw(
			[
				// (center[0]) * DEG2RAD,
				(90 - center[0]) * DEG2RAD,
				center[1] * DEG2RAD,
			],
			R
		)
	}

	project(lng, lat, alt) {
		let lnglatalt = Array.isArray(lng) ? lng.slice() : [lng, lat, alt]

		// lnglatalt[0] = 90 - lnglatalt[0]
		// let _d = lnglatalt[0] - this._center[0]
		// if (_d > 180) {
		// 	_d -= 360
		// }
		// if (_d < -180) {
		// 	_d += 360
		// }
		// console.log(_d);
		// lnglatalt[0] = 90 - (lnglatalt[0] - this._center[0])
		// lnglatalt[0] = 90 - (lnglatalt[0] - this._center[0])
		// lnglatalt[0] = 90 - (lnglatalt[0] - this._center[0])
		// lnglatalt[0] = (lnglatalt[0] - this._center[0])
		// lnglatalt[1] = (lnglatalt[1] - this._center[1])

		// let _d = lnglatalt[1] - this._center[1]
		// if (_d > 90) {
		// 	_d -= 180
		// }
		// if (_d < -90) {
		// 	_d += 180
		// }
		// console.log(lnglatalt[1] - this._center[1]);
		// console.log(_d);
		// lnglatalt[1] = (_d)
		// lnglatalt[2] = (lnglatalt[2] - this._center[2]) || 0

		lnglatalt[2] = lnglatalt[2] || this.getAlt([lnglatalt[0], lnglatalt[1]]) // 如果未指定海拔或者从希望从 heightMap 读取

		lnglatalt[0] = 90 - lnglatalt[0]
		// lnglatalt[1] = lnglatalt[1]

		lnglatalt[0] *= DEG2RAD
		lnglatalt[1] *= DEG2RAD
		const xyz = projRaw(lnglatalt, this.R)

		// debugger

		// return xyz

		// const relativeXYZ = xyz
		// const relativeXYZ = [
		// 	xyz[0],
		// 	xyz[1],
		// 	(xyz[2] || 0) - R,
		// ]
		const relativeXYZ = [
			xyz[0] - this._xyz0[0],
			xyz[1] - this._xyz0[1],
			xyz[2] - this._xyz0[2] || 0,
		]

		// debugger

		return relativeXYZ
	}

	unproject(x, y, z) {
		console.error('接口待实现')
		return [0, 0, 0]
		// let xyz = Array.isArray(x) ? x : [x, y, z]
		//
		// const lnglatalt = unprojRaw(xyz, this.R)
		// lnglatalt[0] /= DEG2RAD
		// lnglatalt[0] = lnglatalt[0]
		// lnglatalt[1] /= DEG2RAD
		// return lnglatalt
	}
}

// console.log('test', new SphereProjection().project([1,0]));

// Mercator公式
function projRaw(lnglatalt, R) {
	const phi = lnglatalt[0]
	const theta = lnglatalt[1]
	const alt = lnglatalt[2] || 0

	const radius = R + alt

	// console.log(phi, theta);

	return [
		radius * Math.cos(phi) * Math.cos(theta),
		radius * Math.sin(theta),
		radius * Math.sin(phi) * Math.cos(theta),
	]
}

// eslint-disable-next-line no-unused-vars
function unprojRaw(xyz, R) {
	const x = xyz[0]
	const y = xyz[1]
	const z = xyz[2]

	const radius = Math.sqrt(x * x + y * y + z * z)
	const alt = radius - R

	const phi = Math.acos(y / radius)
	// const theta = Math.asin(z / radius / Math.sin( phi ))

	// const tanTheta = z / x
	// let theta = Math.atan(z / x)
	// if (theta !== theta) {
	// 	theta = Math.asin(z / radius / Math.sin( phi ))
	// }

	let theta = Math.atan2(y, x)

	// console.log(phi, theta);

	const lng = theta
	const lat = phi

	return [lng, lat, alt]
}

// window.pp = new SphereProjection({
// 	center: [10, 10]
// })

// p = new SphereProjection()

File: packages/utils/projections/csharp/SphereProjection.cs
using System;

namespace Ali.Polaris.Projections
{
    public class SphereProjection : Projection
    {
        private double[] _xyz0 = new double[3];

        public override double[] Center
        {
            get => base.Center;
            set
            {
                base.Center = value;
                var r = ProjRaw(new double[] { (90d - value[0]) * DEG2RAD, value[1] * DEG2RAD }, R);
                _xyz0[0] = r[0];
                _xyz0[1] = r[1];
                _xyz0[2] = r[2];
            }
        }

        public SphereProjection(ProjectionProps props) : base(props)
        {
            IsSphereProjection = true;
            Type = "SphereProjection";
        }

        public override double[] Project(double lng, double lat, double alt = 0d)
        {
            lng = 90d - lng;

            var xy = ProjRaw(new double[] { lng * DEG2RAD, lat * DEG2RAD }, R);

            var x = xy[0] - this._xyz0[0];
            var y = xy[1] - this._xyz0[1];
            var z = xy[2] - this._xyz0[2];


            return TransformOut(x, y, z);
        }

        public override double[] Unproject(double x, double y, double z = 0d)
        {
            TransformIn(ref x, ref y, ref z);
            throw new MissingMethodException("Method not implemented");
        }

        // raw

        protected static double[] ProjRaw(double[] lnglatalt, double R)
        {
            var phi = lnglatalt[0];
            var theta = lnglatalt[1];
            var alt = lnglatalt.Length > 2 ? lnglatalt[2] : 0d;

            var radius = R + alt;

            // console.log(phi, theta);

            return new double[] {
                radius * Math.Cos(phi) * Math.Cos(theta),
                radius * Math.Sin(theta),
                radius * Math.Sin(phi) * Math.Cos(theta)
            };
        }
    }
}